source/common/router/router.cc

```diff
@@ -1426,7 +1426,7 @@ bool Filter::setupRedirect(const Http::ResponseHeaderMap& headers,
       !callbacks_->decodingBuffer() && // Redirects with body not yet supported.
       location != nullptr &&
       convertRequestHeadersForInternalRedirect(*downstream_headers_, *location) &&
-      callbacks_->recreateStream()) {
+      callbacks_->recreateStream(&headers)) {
```

Comment:
I'm curious, how is the internal redirect response reported to stats when the internal redirect fails due to the original request having a body or convertRequestHeadersForInternalRedirect failing?

Is piggybacking on recreateStream the right thing or do we need a new StreamDecoderFilterCallbacks callback to correctly handle stats reporting in this case?

Reply:
When convertRequestHeadersForInternalRedirect fails, setupRedirect will return false back to onUpstreamHeaders, which will continue and eventually call encodeHeaders (which calls chargeStats). The success case couldn't follow this flow because encodeHeaders sends a response back downstream.